lib/SPIRV/SPIRVReader.cpp

```diff
@@ -1739,36 +1739,48 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *BV, Function *F,
   }
 
   case OpFMod: {
-    // translate OpFMod(a, b) to copysign(frem(a, b), b)
+    // translate OpFMod(a, b) to:
+    //   r = frem(a, b)
+    //   c = copysign(r, b)
+    //   needs_fixing = isordered(r, b) && r != c && r != 0
+    //   result = needs_fixing ? r + b : c
+    IRBuilder<> Builder(BB);
     SPIRVFMod *FMod = static_cast<SPIRVFMod *>(BV);
     auto Dividend = transValue(FMod->getDividend(), F, BB);
     auto Divisor = transValue(FMod->getDivisor(), F, BB);
-    auto FRem = BinaryOperator::CreateFRem(Dividend, Divisor, "frem.res", BB);
-
-    std::string UnmangledName = OCLExtOpMap::map(OpenCLLIB::Copysign);
-    std::string MangledName = "copysign";
-
-    std::vector<Type *> ArgTypes;
-    ArgTypes.push_back(FRem->getType());
-    ArgTypes.push_back(Divisor->getType());
-    mangleOpenClBuiltin(UnmangledName, ArgTypes, MangledName);
-
-    auto FT = FunctionType::get(transType(BV->getType()), ArgTypes, false);
-    auto Func =
-        Function::Create(FT, GlobalValue::ExternalLinkage, MangledName, M);
-    Func->setCallingConv(CallingConv::SPIR_FUNC);
-    if (isFuncNoUnwind())
-      Func->addFnAttr(Attribute::NoUnwind);
-
-    std::vector<Value *> Args;
-    Args.push_back(FRem);
-    Args.push_back(Divisor);
-
-    auto Call = CallInst::Create(Func, Args, "copysign", BB);
-    setCallingConv(Call);
-    addFnAttr(Call, Attribute::NoUnwind);
-    return mapValue(BV, Call);
+    auto FRem = Builder.CreateFRem(Dividend, Divisor, "frem.res");
+    auto CopySign = Builder.CreateBinaryIntrinsic(
+        llvm::Intrinsic::copysign, FRem, Divisor, nullptr, "copysign.res");
+    auto Zero = ConstantFP::getNullValue(Dividend->getType());
+    auto CmpSign = Builder.CreateFCmpONE(FRem, CopySign, "cmpsign.res");
+    auto CmpFRem = Builder.CreateFCmpONE(FRem, Zero, "cmpfrem.res");
+    auto FAdd = Builder.CreateFAdd(FRem, Divisor, "fadd.res");
+    auto Cmp = Builder.CreateAnd(CmpSign, CmpFRem, "cmp.res");
+
+    auto Select = Builder.CreateSelect(Cmp, FAdd, CopySign);
+    return mapValue(BV, Select);
+  }
+
+  case OpSMod: {
+    // translate OpSMod(a, b) to:
+    //   r = srem(a, b)
+    //   needs_fixing = a < 0 != b < 0 && r != 0
```

Comment:
Please put some `(`, `)` brackets here, it is a little bit hard to parse this expression

Reply:
Okay, although I'm unsure as to where you want the brackets. Is this okay? :

```
needs_fixing = ((a < 0) != (b < 0) && r != 0)
```

Reply:
I was mostly confused by `a < 0 != b < 0` and spend some time trying to understand this expression :)

Reply:
Ah, yes, `(a ^ b) < 0` is equivalent, but I felt it wouldn't have been helpful to put that in the comment. I'd be happy to reword both of the translation comments if you have any suggestions on how I could improve upon them.